src/librustc/ty/layout.rs

```diff
@@ -1001,12 +1000,12 @@ impl<'a, 'tcx> LayoutDetails {
 
             let mut offset = Size::from_bytes(0);
 
-            if let StructKind::EnumVariant(discr) = kind {
-                offset = discr.size();
+            if let StructKind::EnumVariant(discr_size, union_align) = kind {
                 if !packed {
-                    let discr_align = discr.align(dl);
-                    align = align.max(discr_align);
+                    align = align.max(union_align);
                 }
+                // Round the offset up the union's aggregate alignment.
+                offset = discr_size.abi_align(union_align);
```

Comment:
What I meant was that you wouldn't need special discriminant vs union semantics, just applying a size&align prefix.